---
title: "Downstream_plots"
author: "Rozeena Arif"
date: "2024-07-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r}
library(tidyverse)
library(rtracklayer)
library(data.table)
library(clusterProfiler)
library(org.Hs.eg.db)
library(openxlsx)
library(UpSetR)
library(ggrepel)
library(GenomicRanges)
library(plyranges)
library(ggseqlogo)
library(BSgenome.Hsapiens.UCSC.hg38)
library(universalmotif)
library(ComplexHeatmap)
library(patchwork)

hs <- org.Hs.eg.db

setwd("/home2/2459810a/iCLIP_analysis/NOP2_Innes/")

source("Scripts/iCLIP_analysis_functions_2.R")
source("Scripts/cliprofiler_functions.R")

load("GENOMEDIR/genome_annotations.gtf")
load("GENOMEDIR/genome_annotation_introns.gtf")
```

#  generating PCA plot of counts in each sample
```{r}
# Step 1: Read and Prepare Data
data_matrix <- read.table("/home2/2459810a/iCLIP_analysis/NOP2_Innes/DEW-seq/Matrix_w50s20/Full_NOP2_iCLIP.matrix.txt.gz", header = TRUE, row.names = 1, sep = "\t")
# Optionally, normalize or scale your data heree


# Step 2: Perform PCA
pca_result <- prcomp(t(data_matrix), scale. = TRUE)


# Step 3: Plot PCA
library(ggplot2)

# Create a data frame for plotting
pca_data <- data.frame(Sample = colnames(pca_result$x), PC1 = pca_result$x[,1], PC2 = pca_result$x[,2])

# Assuming you have a vector that maps samples to conditions
 sample_conditions <- c("IP_SINV_1", "IP_SINV_2","IP_SINV_3","IP_mock_1","IP_mock_2","IP_mock_3","SMI_SINV_1","SMI_SINV_2","SMI_SINV_3","SMI_mock_1","SMI_mock_2","SMI_mock_3")
# Make sure the order of sample_conditions matches the columns in data_matrix

# Add conditions to the pca_data
pca_data$Condition <- sample_conditions

# Plot
ggplot(pca_data, aes(x = PC1, y = PC2, color = Condition)) +
  geom_point() +
  theme_minimal() +
  ggtitle("PCA of Sequencing Data") +
  xlab("PC1") + ylab("PC2")
```



#Load files
```{r}
NOP2_files <- list.files(path = "DEW-seq/Regions_w50s20/", pattern = "_enrichedregions.log2FC_2.0_p0.05.bed", full.names = TRUE)
# try with probability of 0.05 and 0.01 both to see the difference. 0.01 is better for highly specific binding sites.
NOP2_files
```

#Convert to granges files and extract binding site information
```{r}
NOP2_bs_gr <- map2(NOP2_files, 
                         c("mock","SINV"),
                         ~ get_gr_and_bs(.x, "NOP2", .y))
```

#Identify gene region binding preference
```{r}
#as counts

NOP2_bs_gr %>% 
map_dfr(~ .x[["bs"]]) %>% 
  #dplyr::filter(gene_biotype == "protein_coding") %>% 
  dplyr::filter(type %in% c("five_prime_utr", "CDS", "three_prime_utr", "intron")) %>% 
  distinct(id, type, condition) %>% 
  group_by(condition) %>% 
  count(type) %>%
  #dplyr::mutate(percent = n*100/sum(n)) %>% 
  dplyr::mutate(condition = factor(condition, levels = c("mock","SINV"))) %>% 
  ggplot(aes(x = condition, y = n, fill = type)) + geom_col(position = "dodge",colour = "black") +
  theme_bw()

#as proportions
NOP2_bs_gr %>% 
map_dfr(~ .x[["bs"]]) %>% 
  dplyr::filter(type %in% c("five_prime_utr", "CDS", "three_prime_utr", "intron")) %>% 
  distinct(id, type, condition) %>% 
  group_by(condition) %>% 
  count(type) %>%
  dplyr::mutate(percent = n*100/sum(n)) %>% 
  dplyr::mutate(condition = factor(condition, levels = c("mock","SINV"))) %>% 
  ggplot(aes(x = condition, y = percent, fill = type)) + geom_col(colour = "black") +
  theme_bw()
```

#Generate meta-gene profile of all binding sites

This step requires an additional genome annotation and functions from the cliProfiler package.
```{r}
anno <- rtracklayer::import.gff3(con = "GENOMEDIR/gencode.v41.annotation.gff3.gz")

metagenes <- NOP2_bs_gr %>% map(~ metaGeneProfile(.x[["gr"]], anno, split = FALSE))

mock_meta <- metagenes[[1]]$Plot + theme(legend.position = "none") + ggtitle("mock")
sinv_meta <- metagenes[[2]]$Plot + theme(legend.position = "none")+ ggtitle("SINV")

mock_meta + sinv_meta
```

#Gene biotype preference
```{r}

map_dfr(NOP2_bs_gr, function(x){
  
pseudo <- grep("pseudogene", x[["bs"]]$gene_biotype, value = TRUE)
  
x[["bs"]] %>% 
  dplyr::mutate(gene_biotype = ifelse(gene_biotype %in% pseudo, "pseudogene", gene_biotype)) %>% 
  distinct(id, gene_biotype, condition) %>% 
  group_by(condition) %>% 
  count(gene_biotype) %>%
  dplyr::mutate(percent = n*100/sum(n)) %>% 
  dplyr::arrange(percent) %>% 
  dplyr::mutate(gene_biotype = factor(gene_biotype, levels = gene_biotype))

}) %>% 
  dplyr::mutate(condition = factor(condition, levels = c("mock", "SINV"))) %>% 
  ggplot(aes(x = gene_biotype, y = percent, fill = gene_biotype)) + 
  geom_col(colour = "black") + coord_flip() + facet_wrap(~condition) + theme_bw() +
  theme(legend.position = "none") 
```

#Information on number of binding sites and target genes
```{r}
count_info <- NOP2_bs_gr %>% map_dfr(~ get_bs_count_info(.x[["bs"]], .x[["gr"]]))

count_info %>% 
  dplyr::filter(type != "n_unique_binding_sites") %>% 
  dplyr::mutate(type = ifelse(type == "n_unique_binding_sites_nonoverlapping", "binding sites",
                              "target genes")) %>% 
  dplyr::mutate(condition = factor(condition, levels = c("mock", "SINV"))) %>% 
  ggplot(aes(x = condition, y = count, fill = type)) + 
  geom_col(position = "dodge",colour = "black") +
  theme_bw() + scale_fill_manual(values = c("#6F1B45", "#1B6F1B"))
```

#Upset plot of overlaps between different groups
```{r}
gene_lists <- NOP2_bs_gr %>% 
  map(function(x){
    
    x[["bs"]] %>% pull(gene_name) %>% unique()
    
  }) 

names(gene_lists) <- c("mock", "SINV")
  
  
upset(fromList(gene_lists), order.by = "freq")
```

#Venn diagram of overlaps between different samples
A venn diagram is better when there are fewer comparisons to make
```{r}

# pdf("overlap_venn.pdf", width = 1.7, height = 3)
plot(eulerr::euler(gene_lists, shape = "ellipse"), 
     quantities = list(cex = 1),
     fills = c("#6a3d9a", "#1f78b4", "#489FA7"))
# dev.off()

```

#GO enrichment analysis for target genes
```{r}
#the organism database is loaded at the top as a package would be
e_ids_NOP2_mock_bound <- AnnotationDbi::select(hs, 
       keys = gene_lists[[2]],
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL") %>% pull(ENTREZID)

#BP (biological process) can be changed to MF (molecular function) or CC (cellular compartment)
ego_NOP2_mock_bound <- enrichGO(
                gene          = e_ids_NOP2_mock_bound,
                #universe      = names(geneList),
                OrgDb         = hs,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable      = TRUE)



#many GO terms are overlapping so this function seeks to combine overlapping terms and simplify the output
#the cutoff can be adjusted, depending on how many GO terms you want/ how similar they are
ego_NOP2_mock_bound_simp <- simplify(ego_NOP2_mock_bound, cutoff=0.6, by="p.adjust", select_fun=min)

ego_NOP2_mock_bound_simp %>% as.tibble()%>% 
  tidyr::separate(GeneRatio, c("gene_go", "gene_no_go"))%>% 
  tidyr::separate(BgRatio, c("back_go", "back_no_go")) %>% 
  dplyr::mutate_at(c("gene_go", "gene_no_go", "back_go", "back_no_go"), ~ as.numeric(.)) %>% 
  dplyr::mutate(gene_no_go = gene_no_go - gene_go, back_no_go = back_no_go - back_go) %>% 
  dplyr::mutate(odds_ratio = (gene_go/gene_no_go)/(back_go/back_no_go)) %>% 
  ggplot() + geom_point(aes(x = Description, y = "1", size = log2(odds_ratio), colour = -log10(p.adjust))) + coord_flip() + theme_bw() + theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.title.y=element_blank(),
        axis.ticks.x=element_blank(),
         #plot.margin=grid::unit(c(0,0,0,0), "mm")#, 
        #legend.position = "none"
        ) + 
  scale_colour_gradient2(
    low = "#f2f0f7", 
    mid = "#9e9ac8", 
    high = "#54278f",
    midpoint = 1e-06
  )

```


#Plot coverage of signal across the viral genome
```{r}
SINV_coverage <- read.csv("GENOMEDIR/NOP2_SINV_coverage.csv")

SINV_coverage_means <- SINV_coverage %>% 
  group_by(condition, X2, X3, sample) %>% dplyr::summarise(mean = mean(X4)) 

SINV_coverage_means %>% 
  dplyr::filter(sample == "SINV") %>% 
  pivot_wider(names_from = condition, values_from = mean) %>% 
  dplyr::mutate(SMI = ifelse(is.na(SMI), 0, SMI )) %>% 
  dplyr::mutate(normalised = IP - SMI) %>%
  dplyr::mutate(direction = ifelse(normalised < 0, "SMI", "IP")) %>% 
   ggplot(aes(x = X3, y = normalised, colour = direction)) +
  geom_line( size = 1) + scale_colour_manual(values = c("#2c7fb8", "grey80"))
```


